### DISCUSSION TITLE
How to set the "lang" attribute programatically? In a +page.server.ts for example

### DISCUSSION THREAD

**Discussion by Antoine-lb** [sveltejs/kit#12376] [Q&A] [answered]
### Describe the problem

Hello 👋
If I have a blog where each article can be in a different language, the [recomended version](https://kit.svelte.dev/docs/accessibility#the-lang-attribute) on how to handle the `lang` attribute falls short.

This is because I only know what language the page is when I load the article (in the `+page.server.ts`), but the `transformPageChunk` already returned.

This problem is very annoying bc if not done right, on some Androids Google will just translate the site without asking (is a opt out) just based on this `lang` attribute. And Google translate is very bad at translating.

### Describe the proposed solution

Using `<svelte:head>` to modify the `<html>` tag. Like a `<svelte:html>`.

### Alternatives considered

Run `transformPageChunk` in a load function.

### Importance

would make my life easier

### Additional Information

Other people with the same problem:
- https://github.com/sveltejs/kit/issues/5508
- https://github.com/sveltejs/kit/issues/3091#issuecomment-1541880297

**Comment by Antoine-lb** [comment]
Another potential solution would be to add a `%sveltekit.lang%` the same way as `%sveltekit.head%` or `%sveltekit.body%`

**Comment by dominikg** [comment]
you posted here about the same problem already. https://github.com/opral/inlang-paraglide-js/issues/134

converting to discussion as this is not an issue with sveltekit itself but rather a question about an unusual arcitecture.

**Reply by Antoine-lb** [reply]
Hello 👋
I disagree. It's not the same problem, what I posted there is related to Paraglide way of handeling urls. `sveltekit-i18n` handles my use case pretty well so I will stick with them.

I do believe that is an issue with SvelteKit architecture, with the current tools you *have* to know in advance what language a page is.

I can't change the `lang` tag on a `+page.server.ts`, even thought we are still on the server, so we still have time to do things properly.

On a PHP server I can do this very easily, but on SvelteKit I have to do it on the client side.

I know it feels like a unusual architecture, but in reality we are dealing with user generated content, that is why we need the database to tell us what language it is.

**Comment by dominikg** [comment]
if you have no information available in the url directly about the language to be used, you can create a lookup table and access that in the handle hook. 

Remember to also translate the whole page to that language if you set lang at the html level, not just the article itself.

If the page is always in one language, and the article in another, you can use embedded lang attribute

```
<html lang="en">
...english ui here, menu. headers, you name it ...
  <article lang="de">
  ...  hier steht ein eingebetteter deutscher Artikel ...
  </article>
... back to english for footer or whatever ...  
</html>
```

**Reply by Antoine-lb** [reply]
> create a lookup table and access that in the handle hook.

By this you mean that I should fetch all articles in the hook to know language each of them are?

> If the page is always in one language, and the article in another, you can use embedded lang attribute

Could be a nice solution, will depend on how google handles it bc currently is constantly asking users to translate, is why I got into this mess in the first place

Thanks for taking the time to answer my questions 🙏

**Comment by david-plugge** [comment]
You can use `transformPageChunk`. Set the locale/lang in your load functions to the request locals and ready that value in the handle Hook. As locals should be passed as reference it should work.

**Reply by Antoine-lb** [reply]
That looks like it could work but I don't understand all the steps involved. How can I run `transformPageChunk` after the load function?

**Reply by jhubbardsf** [reply]
Check this out here in the docs for [hooks.server.ts](https://kit.svelte.dev/docs/hooks#server-hooks). This example from that page might help you. As far as I know, during SSR hooks runs after load. On Client side navigation load runs, then on client side navigation your load will run before hooks. When you hit a .server.ts I believe hooks.ts will run.

```svelte
import type { Handle } from '@sveltejs/kit';

export const handle: Handle = async ({ event, resolve }) => {	
   const response = await resolve(event, {		
   transformPageChunk: ({ html }) => html.replace('old', 'new'),		
   filterSerializedResponseHeaders: (name) => name.startsWith('x-'),		
   preload: ({ type, path }) => type === 'js' || path.includes('/important/'),	
});
return response;
};

**Comment by Antoine-lb** [comment, marked as answer]
Here is the solution:
`+page.server.ts`
```js
export const load = (async (event) => {
        // use event.locals to set the value to be passed to the hook
	event.locals.locale = 'fr';

	return {};
});
```

`hook.server.ts`
```js
export async function handle({ event, resolve }) {
	return resolve(event, {
                // then you can read this local here bc SvelteKit will run this function after the +page.server.ts
		transformPageChunk: ({ html }) => html.replace('%lang%', event.locals.locale) 
	});
}
```

Thanks @david-plugge and @jhubbardsf for the solution.
